### DISCUSSION TITLE
NewUI - the third

### DISCUSSION THREAD

**Discussion by dcorson-ticino-com** [jamulussoftware/jamulus#1343] [UI Related]
I finally have something that I, at least, like.
Tabbed docking windows so one can make a big window multi-paned UI or undock all the windows and put them wherever you want on the screen.

https://github.com/dcorson-ticino-com/jamulus/releases/tag/r3_8_0newUIbeta3

I have gotten so used to this UI that I now have problems going back to the standard Jamulus UI.
At program start the Connect window is open.  After connecting to a server the Connect window is hidden and the chat window is raised.
If the Connect window is hidden when one clicks on Connect it is raised.  Double clicking a server at any time closes the presently connected server and connects to the new one.  To reduce the ping traffic, after the Connect window is raised by any means the servers are pinged for 20 seconds (8 times) then the pinging is reduced to once a minute.  Update restarts the 20 second fast pinging.  The window can be raised by clicking connect, clicking disconnect or clicking the tab

![NewUI03a](https://user-images.githubusercontent.com/64357360/112391263-eca55100-8cf7-11eb-9e1c-f98cb9ba41b7.JPG)

Windows can be docked left and right
![NewUI03b](https://user-images.githubusercontent.com/64357360/112391272-f0d16e80-8cf7-11eb-9367-5e860e06a3d9.JPG)

Or undocked
![NewUI03c](https://user-images.githubusercontent.com/64357360/112391280-f3cc5f00-8cf7-11eb-837f-28c900035f61.JPG)

![NewUI03d](https://user-images.githubusercontent.com/64357360/112391303-fb8c0380-8cf7-11eb-82b6-4474ed6f1ef5.JPG)

![NewUI03e](https://user-images.githubusercontent.com/64357360/112391310-fdee5d80-8cf7-11eb-8a9a-490a5a9475e1.JPG)

![NewUI03f](https://user-images.githubusercontent.com/64357360/112391325-03e43e80-8cf8-11eb-98e0-f68db32794f8.JPG)

![NewUI03g](https://user-images.githubusercontent.com/64357360/112391357-13638780-8cf8-11eb-8f53-01d21d9b0a7e.JPG)

![NewUI03h](https://user-images.githubusercontent.com/64357360/112391376-1bbbc280-8cf8-11eb-9aba-b9db73c2412c.JPG)

What do you think ?
This is now my suggestion if it should be decided to go forward with a new UI for Jamulus.

**Comment by ann0see** [comment]
For me, it’s not really intuitive to click the connect button on the mixer board if I have to select the server on the left. Would it be possible to move the connect button somewhere else or the server connect dialogue into the mixer board?

Also I‘m a bit afraid to include it at the moment, since it might make jamulus look more complicated (although in reality it does the contrary). I‘d prefer to present it quite similar to the first time user. But that’s for an open discussion.

**Reply by dcorson-ticino-com** [reply]
Of course that is possible, but it means a button more on the screen.
I have one button doing 3 things that can't get mixed up, raise window, connect & disconnect.
The button is always at the same place no matter where the connect window may happen to be on the screen.
And we don't have the problem we have now that there are 2 connect buttons shown and one does nothing at times.

**Reply by ann0see** [reply]
Probably true, but it's not very intuitive for private servers to have a connect button somewhere across the screen.

**Reply by dcorson-ticino-com** [reply]
OK, I see what you mean.  
I don't know how others use Jamulus, but I have 2 use cases.
1) rehearsals on a known server, be it public or private.  
That server was entered once in the Server Address field and stays there from use to use.
To use it I click the connect button without even going to the Connect window as when nothing is selected in the list the server address field is used.
2) looking at what is happening in the world, jamming.
A double-click on the server to be selected does the job, Connect button not needed.

I suggest, as before, to download and try it out in use.  
There are cases where you may go back and forth, but I have not seen that happening often. 
No matter where you put the button there are cases where it is not the perfect placing for the actions that follow.

**Reply by ann0see** [reply]
Agree. I did try it and that's what came to my mind. First time (private server) users *will* be confused by this. At least that's what I think. The first contact to Jamulus ideally should work like this:

1. Download Jamulus
2. Install Jamulus
3. Click on connect or enter a server
4. Start Jamming

**Reply by ann0see** [reply]
Let's agree on a compromise:

If you input an address and press the return key --> jamulus should connect to this server.

**Reply by dcorson-ticino-com** [reply]
>If you input an address and press the return key --> jamulus should connect to this server.

Should Return in general be a shortcut for the connect button ?  disconnect too ?  in the conncet window ? the main window ?

**Reply by ann0see** [reply]
Return in the connect window if your cursor is in the server adress field --> connect to this server

Return in the main window --> nothing

Return if you selected a server (no double click) --> connect to the selected server

**Comment by ann0see** [comment]
Ok. Now a more detailed feedback:

1. Having the tabbed settings is great! Especially that the profile is in settings now (where I'd probably expect it to be). 
2. Deviding audio and network setup probably has reasons. Buffer size is something related to Audio while network buffer isn't. So in theory network settings is the wrong place. I'm unsure if we really want to devide audio and network setup? Maybe we could introduce a basic setup page which shows: 

Inputs/outputs
ASIO-Device settings
buffer (both network and audio buffers)). 

An advanced page which shows

Central server
measurements
audio quality
input balance

etc. 

I'd be in favour of implementing some of your improvements in the next version, but not the whole UI yet.

This allows users to slowly adapt to a new UI.

**Reply by dcorson-ticino-com** [reply]
To tell you the truth, the split audio, network is only to keep the windows small and vertical so they fit in the same space as the connect list. 
No other reason.  
We can mix and match at will.

**Comment by ngocdh** [comment]
Thanks @dcorson-ticino-com for this new UI. I'm loving it. It definitely requires some getting used to, but it's much better than separate windows. 

I have a few observations here:
- On Windows, I see shortcuts for "My Profile" and "Audio Setup", very convenient. Could you also add shortcut for "Network setup" as well, because musicians might need to quickly adjust their network settings while jamming.
- Chat window stays silent when a message is received. Expected behavior: Chat tab changes color, or add number of pending messages in parentheses: Chat (1)
- When chatting, typing enter should send the message. It's now doing nothing
- I see great improvements in the UI of the Android app. However, Settings and Connect tabs are not using the whole width of the panel, in contrary to Chat and Actions tabs. Also, shortcuts in the context menu are not showing / working.

**Reply by ngocdh** [reply]
After playing with the new UI a while, I think the Settings tab should be removed, instead the sub-tabs "My Profile", "Audio Setup" and "Network Setup" should be upgraded to main tabs for easier access.

**Reply by dcorson-ticino-com** [reply]
Thank you for your comments @ngocdh.  These are exactly the kinds of improvement propositions I am looking for.
It is interesting to hear about the Android usage case as I have no experience or access to that myself.

The settings tab with its sub-tabs is implemented that way as this is all the same object at present and it was quicker this way.  That doesn't mean it must stay that way 'though.

Please realize this is a quick hack to see if this is the direction we should go in or not.  It is not a completely developed new UI.  Things like shortcuts and easy access are implemented only randomly.  This is my 5th or 6th try at an improvement in the UI, the 3rd that I have published here.
Personally I am very happy going this direction, but I am not alone here.

**Reply by ngocdh** [reply]
IMO the direction is quite good. Keep up the good work :)

**Comment by ngocdh** [comment]
@dcorson-ticino-com, I wanted to test your UI on Mac, but it seems the releases were removed. Can I expect to see them back online?

Cheers,

**Reply by dcorson-ticino-com** [reply]
I am surprised, I did not remove them.  I will check.

EDIT: I see what happened, but I don't have time to repair right now.  Give me a day or two.

**Reply by henkdegroot** [reply]
@dcorson-ticino-com I would love to test as well. It looks amazing, with all kinds of fancy features. Will wait for the release to back online :) Excellent Work!

**Reply by dcorson-ticino-com** [reply]
It is now again possible to download from here:
https://github.com/dcorson-ticino-com/jamulus/releases/tag/r3_8_0beta3NewUI

Let me know what you think please.

**Reply by ngocdh** [reply]
Thanks @dcorson-ticino-com. I was able to test it on Mac. A bit weird that the tab name is blank whenever the tab is active, a bug maybe?
Also, I would love to be able to set the default active window ("Audio setup" for example) instead of the "Connect" window. Either using the command line, or a new attribute in settings.cpp?

**Reply by dcorson-ticino-com** [reply]
Thanks for testing, unfortunately I don't have access to a MAC to debug.  But when this gets more serious and potentially fit for a release we will get it.
What do you mean by default window?  Right now the top window is Connect when not connected and Chat when connected as I think those are the windows that will be most used in those conditions.  When should the "default" window be shown ?  I am trying to avoid having many settings that control special cases as that makes the interface complicated and difficult to understand. 
Have you seen that you can grab the windows by the title bar and undock them and place them to the left or right of the main window and also anywhere on the screen?

**Reply by ngocdh** [reply]
> Thanks for testing, unfortunately I don't have access to a MAC to debug. But when this gets more serious and potentially fit for a release we will get it.

=> Agreed!

> What do you mean by default window? Right now the top window is Connect when not connected and Chat when connected as I think those are the windows that will be most used in those conditions. When should the "default" window be shown ? I am trying to avoid having many settings that control special cases as that makes the interface complicated and difficult to understand.

=> Yes I also like to keep the interface simple, which is why I suggested the command line or ini file options. My use case is that, I open Jamulus client using command line and ini settings, which connects to a hardcoded server, so I don't really need the Connect window. Having "Settings" instead of "Connect" is more useful for me. (workaround found, more on that below)

> Have you seen that you can grab the windows by the title bar and undock them and place them to the left or right of the main window and also anywhere on the screen?

=> Yes I have. For now all my windows are docked. And thanks to you I found a workaround for my need: I'll just put the settings tab to the right, so that I have Connect and Settings on the interface at the same time. One on the left, one on the right. :)

**Reply by henkdegroot** [reply]
Thank you for the effort in making this available for testing. Really helps a lot, as sometimes you only see/find things when you can use it yourself.
Here some thoughts:

When I started the first time, the window was very small (most likely due to the fact I used 3.7.0 with a small size window). Guess that will be fixed in a final version.
Also accessing the 

Also, again related to 3.7.0 usages, the last session I had the settings dialog open. So the new version shows the settings dialog undocked. However I could not see all settings, I had to increase the window size to see all options. Adding a scroll bar could help or making sure a minimum size is set. Again, most likely something to fix in a final version when thinking about migration users and how they will experience the new layout.

**Settings tab/dialog:**
My Profile tab, this is showing more then just my profile. Sounds more like these are User Preferences/settings. I think it would better to have this tab as the last tab in the dialog as you will not use it very often. 

Perhaps some control would be good of which tab should be "focussed" when the dialog is shown? I.e. using a setting to let the user decide or decide what function is most likely the user wants to access when they open/select the settings dialog.
Or should there even be "sub-tabs" in the dialog? With the dock/undock features, perhaps the tabs in settings should just be "main" tabs? This will also allow someone to have all tabs visible (when undocked) at once.

**Actions tab/dialog:**
Is it really need to have this? Some items would seem more appropiate to move into the tab which also is used to select the skin, etc. 

**Dock/undocking:**
It might not be clear for first time users that you can dock/undock. A hover (tooltip) could help to indicate this. Sometimes the title bar looks like a button and users may want to click on it. If then the tooltip is displayed, it should be clear what it can be used for.

**Main dialog:**
Does not include a menu item for Network settings (yet I guess)....

All the above just thoughts....hope some will be of help.

**Reply by henkdegroot** [reply]
> Thanks @dcorson-ticino-com. I was able to test it on Mac. A bit weird that the tab name is blank whenever the tab is active, a bug maybe?

I have checked on a mac as well and do have text in the tab, even when it is active. Compared to the windows version, there is just no menu at the top of the main dialog.

**Comment by ngocdh** [comment]
No rush :)
